File: src/main/utils/command.ts
// PX Dev — Command Builder
// 结构化命令解析 + Windows .cmd 重映射 + 批处理脚本 shell 兼容

import { spawn } from 'child_process'
import type { Service } from '@shared/types'
import { WINDOWS_CMD_REMAP } from '@shared/constants/defaults'
import { logger } from './logger'

export interface BuiltCommand {
  executable: string
  args: string[]
  options: {
    cwd: string
    env: NodeJS.ProcessEnv
    shell: boolean
    windowsHide: boolean
  }
}

/** Windows 批处理脚本扩展名 */
const WINDOWS_BATCH_EXT = /\.(cmd|bat)$/i

/**
 * cmd.exe 下必须加引号才能安全传递的字符。
 *
 * 包含空白字符与 shell 元字符：一旦命中，就用双引号包裹整个实参，
 * 否则 cmd.exe 会把它当成多个 token 或语法结构（重定向 / 管道 / 转义）来解析。
 */
const CMD_SPECIAL_CHARS = /[\s&|<>^"()%!,;=]/

/**
 * 判断可执行文件是否为 Windows 批处理脚本（.cmd / .bat）。
 *
 * 自 Node 18.20.2 / 20.12.2 起（CVE-2024-27980 的缓解措施），
 * `child_process.spawn()` 在 **shell 为 false** 时拒绝直接执行 .cmd / .bat，
 * 并**同步抛出** `Error: spawn EINVAL`。因此这类脚本必须走 shell:true。
 *
 * @param executable 已完成重映射的可执行文件名或路径
 * @param platform   目标平台，默认当前平台（显式传入便于跨平台单测）
 */
export function isWindowsBatchScript(
  executable: string,
  platform: NodeJS.Platform = process.platform,
): boolean {
  return platform === 'win32' && WINDOWS_BATCH_EXT.test(executable)
}

/**
 * 为 cmd.exe 转义单个 token（可执行文件名或实参）。
 *
 * 背景：当 spawn 使用 shell:true 时，Node 只是把 `[file, ...args]` 用空格拼成一整行
 * 交给 `cmd.exe /d /s /c "<整行>"`，**不会**为任何实参补引号。
 * 所以带空格的路径（如 `C:\Program Files\...`）或带元字符的参数必须由我们自己包裹，
 * 否则会被 cmd 拆散成多个参数，导致命令行语义被破坏。
 *
 * 规则：
 * - 空串 → `""`（保留「显式空参数」语义，不能被 cmd 吞掉）
 * - 不含特殊字符 → 原样返回（保持命令行可读性）
 * - 其余 → 双引号包裹，内部的 `"` 转义为 `\"`（MSVCRT 命令行解析约定）
 *
 * 注意：cmd.exe 的 `%VAR%` 在双引号内仍会展开，这是 cmd 的固有行为，
 * 无法在命令行层面彻底规避；PX Dev 的命令来源是用户自己的服务配置，风险可控。
 */
export function quoteForCmd(value: string): string {
  if (value === '') {
    return '""'
  }
  if (!CMD_SPECIAL_CHARS.test(value)) {
    return value
  }
  return `"${value.replace(/"/g, '\\"')}"`
}

/**
 * 从 Service 定义构建结构化命令。
 *
 * - structured 模式（默认）：command 与 args 分离，Windows 下按表重映射为 .cmd
 * - shell 模式：shell:true，command 原样交给 shell 解析
 *
 * ⚠️ Windows 关键修复：structured 模式下若重映射结果是 .cmd / .bat（例如
 * `npm` → `npm.cmd`），必须强制切换到 shell:true，否则新版 Node 会直接抛
 * `spawn EINVAL`。切换后同步对可执行文件与全部实参做 cmd 转义。
 *
 * 关于「shell 模式下 args 被拼进 executable、返回的 args 为空数组」：
 * Node 在 shell:true 时内部就是执行 `[file, ...args].join(' ')` 再交给 cmd.exe，
 * 且**不会**转义任何实参（Node 22+ 为此新增了 DEP0190 弃用警告）。
 * 既然转义必须由我们自己完成，就干脆把命令行一次拼好：
 * 语义与 Node 的内部拼接完全等价，同时避免触发该弃用警告。
 *
 * @param service  服务定义（command、args、cwd、env）
 * @param platform 目标平台，默认当前平台（显式传入便于跨平台单测）
 * @returns 可直接交给 child_process.spawn() 的参数三元组
 * @throws Error 当启动命令为空时
 */
export function buildCommand(
  service: Service,
  platform: NodeJS.Platform = process.platform,
): BuiltCommand {
  const command = typeof service.command === 'string' ? service.command.trim() : ''
  if (command === '') {
    throw new Error('启动命令为空：请在服务配置中填写「启动命令」')
  }

  const cwd = service.cwd
  const env: NodeJS.ProcessEnv = { ...process.env, ...service.env }
  const shellMode = service.shellMode ?? false

  // 防御性过滤：配置文件损坏时 args 可能混入非字符串元素，spawn 会因此抛错
  let args: string[] = (service.args ?? []).filter(
    (item): item is string => typeof item === 'string',
  )
  let executable: string
  let shell = shellMode

  if (shellMode) {
    // Shell 模式：用户显式选择由 shell 解析整条命令，此处不改写命令内容本身，
    // 只是把实参提前拼进命令行（与 Node 内部行为一致，见下方说明）。
    executable = args.length > 0 ? [command, ...args].join(' ') : command
    args = []
  } else {
    // 结构化模式：拆分可执行文件与实参，并应用 Windows .cmd 重映射
    executable = resolveExecutable(command, platform)

    if (isWindowsBatchScript(executable, platform)) {
      // 批处理脚本无法在 shell:false 下 spawn（EINVAL），改走 cmd.exe 并自行转义
      shell = true
      executable = [quoteForCmd(executable), ...args.map(quoteForCmd)].join(' ')
      args = []
    }
  }

  logger.debug(
    `buildCommand: executable=${executable}, args=${JSON.stringify(args)}, shell=${shell}`,
  )

  return {
    executable,
    args,
    options: {
      cwd,
      env,
      shell,
      windowsHide: true,
    },
  }
}

/**
 * 解析可执行文件名，应用 Windows .cmd 重映射。
 * Windows 下 npm / pnpm / yarn 等实际是 .cmd 包装脚本，需要补后缀。
 *
 * @param command  用户配置的命令名
 * @param platform 目标平台
 */
export function resolveExecutable(command: string, platform: NodeJS.Platform): string {
  // 仅在 Windows 上重映射
  if (platform === 'win32') {
    const lower = command.toLowerCase()
    // 已带扩展名的命令不再重映射
    const hasExt = /\.(exe|cmd|bat|ps1)$/i.test(command)
    if (!hasExt) {
      const remapped = WINDOWS_CMD_REMAP[lower]
      if (remapped) {
        return remapped
      }
    }
  }
  return command
}

/**
 * 用 BuiltCommand 启动子进程。
 * child_process.spawn 的轻量包装，保证调用方式统一。
 */
export function spawnCommand(built: BuiltCommand) {
  return spawn(built.executable, built.args, built.options)
}

File: src-tauri/src/process/command.rs
use crate::types::Service;

/// Built command ready for spawning
pub struct BuiltCommand {
    pub program: String,
    pub args: Vec<String>,
    pub use_shell: bool,
}

/// Build platform-specific command from Service config
pub fn build_command(service: &Service) -> BuiltCommand {
    let command = service.command.trim();
    
    // Check if explicit shell mode
    if service.shell_mode.unwrap_or(false) {
        return build_shell_command(command, &service.args);
    }
    
    // Auto-detect Windows batch wrappers
    #[cfg(target_os = "windows")]
    {
        if should_use_shell_on_windows(command) {
            return build_shell_command(command, &service.args);
        }
    }
    
    // Structured spawn (no shell)
    BuiltCommand {
        program: command.to_string(),
        args: service.args.clone().unwrap_or_default(),
        use_shell: false,
    }
}

/// Check if command should use shell on Windows
#[cfg(target_os = "windows")]
fn should_use_shell_on_windows(command: &str) -> bool {
    let cmd_lower = command.to_lowercase();
    
    // Check if already has batch extension
    if cmd_lower.ends_with(".cmd") || cmd_lower.ends_with(".bat") {
        return true;
    }
    
    // Known batch wrappers that require cmd.exe
    matches!(
        cmd_lower.as_str(),
        "npm" | "pnpm" | "yarn" | "npx" | "bun" | 
        "tsx" | "mvn" | "mvnw" | "gradle" | "gradlew"
    )
}

/// Resolve command to actual executable with extension
#[cfg(target_os = "windows")]
fn resolve_windows_executable(command: &str) -> String {
    let cmd_lower = command.to_lowercase();
    
    // Already has extension, use as-is
    if cmd_lower.ends_with(".cmd") || cmd_lower.ends_with(".bat") || cmd_lower.ends_with(".exe") {
        return command.to_string();
    }
    
    // Map known tools to their actual executables
    match cmd_lower.as_str() {
        "npm" | "pnpm" | "yarn" | "npx" | "tsx" | "mvn" | "mvnw" => {
            format!("{}.cmd", command)
        }
        "gradle" | "gradlew" => {
            format!("{}.bat", command)
        }
        "bun" => {
            format!("{}.exe", command)
        }
        _ => command.to_string()
    }
}

/// Build shell-wrapped command
fn build_shell_command(command: &str, args: &Option<Vec<String>>) -> BuiltCommand {
    #[cfg(target_os = "windows")]
    {
        build_windows_shell_command(command, args)
    }
    
    #[cfg(not(target_os = "windows"))]
    {
        build_unix_shell_command(command, args)
    }
}

/// Quote argument for Windows cmd.exe
#[cfg(target_os = "windows")]
fn quote_for_cmd(arg: &str) -> String {
    // Empty string needs quotes
    if arg.is_empty() {
        return "\"\"".to_string();
    }
    
    // Check if needs quoting
    let needs_quote = arg.contains(' ') || arg.contains('\t') || 
                      arg.contains('&') || arg.contains('|') || 
                      arg.contains('<') || arg.contains('>') || 
                      arg.contains('^') || arg.contains('\"') ||
                      arg.contains('(') || arg.contains(')') ||
                      arg.contains('%') || arg.contains('!') ||
                      arg.contains(',') || arg.contains(';') ||
                      arg.contains('=');
    
    if !needs_quote {
        return arg.to_string();
    }
    
    // Quote and escape
    let mut result = String::from("\"");
    for ch in arg.chars() {
        match ch {
            '\"' => result.push_str("\\\""),
            '\\' => {
                // Check if backslash is before quote
                result.push('\\');
            }
            _ => result.push(ch),
        }
    }
    result.push('\"');
    result
}

/// Build Windows cmd.exe wrapped command
#[cfg(target_os = "windows")]
fn build_windows_shell_command(command: &str, args: &Option<Vec<String>>) -> BuiltCommand {
    // Resolve command to actual executable
    let resolved_cmd = resolve_windows_executable(command);
    
    // Quote the command itself if it has spaces
    let quoted_cmd = if resolved_cmd.contains(' ') {
        format!("\"{}\"", resolved_cmd)
    } else {
        resolved_cmd
    };
    
    // Build full command line with quoted args
    let mut full_cmd = quoted_cmd;
    if let Some(args) = args {
        for arg in args {
            full_cmd.push(' ');
            full_cmd.push_str(&quote_for_cmd(arg));
        }
    }
    
    // Wrap with cmd.exe /d /s /c
    BuiltCommand {
        program: "cmd".to_string(),
        args: vec![
            "/d".to_string(),  // Disable AutoRun
            "/s".to_string(),  // Strip quotes
            "/c".to_string(),  // Execute and exit
            full_cmd,
        ],
        use_shell: true,
    }
}

/// Build Unix shell wrapped command
#[cfg(not(target_os = "windows"))]
fn build_unix_shell_command(command: &str, args: &Option<Vec<String>>) -> BuiltCommand {
    let mut full_cmd = command.to_string();
    if let Some(args) = args {
        for arg in args {
            full_cmd.push(' ');
            full_cmd.push_str(arg);
        }
    }
    
    BuiltCommand {
        program: "sh".to_string(),
        args: vec!["-c".to_string(), full_cmd],
        use_shell: true,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::{Service, ServiceType, ServiceRole};

    fn create_test_service(command: &str, args: Option<Vec<String>>, shell_mode: Option<bool>) -> Service {
        Service {
            id: "test".to_string(),
            workspace_id: "ws".to_string(),
            name: "Test".to_string(),
            service_type: ServiceType::Node,
            role: ServiceRole::Backend,
            cwd: "/path".to_string(),
            command: command.to_string(),
            args,
            package_manager: None,
            port: None,
            env: None,
            env_file: None,
            enabled: true,
            dependencies: vec![],
            startup_delay: None,
            auto_open_browser: None,
            open_url: None,
            health_check: None,
            shell_mode,
            discovery: None,
            created_at: "2024-01-01T00:00:00Z".to_string(),
            updated_at: "2024-01-01T00:00:00Z".to_string(),
        }
    }

    #[test]
    #[cfg(target_os = "windows")]
    fn test_npm_uses_shell_on_windows() {
        let service = create_test_service("npm", Some(vec!["run".to_string(), "dev".to_string()]), None);
        let built = build_command(&service);
        
        assert_eq!(built.program, "cmd");
        assert_eq!(built.use_shell, true);
        assert!(built.args.contains(&"/c".to_string()));
        assert!(built.args.last().unwrap().contains("npm.cmd"));
    }

    #[test]
    fn test_explicit_shell_mode() {
        let service = create_test_service("echo", Some(vec!["hello".to_string()]), Some(true));
        let built = build_command(&service);
        
        assert_eq!(built.use_shell, true);
    }

    #[test]
    fn test_structured_spawn_for_exe() {
        let service = create_test_service("node", Some(vec!["index.js".to_string()]), Some(false));
        let built = build_command(&service);
        
        assert_eq!(built.program, "node");
        assert_eq!(built.use_shell, false);
        assert_eq!(built.args, vec!["index.js"]);
    }
}
